python/triton/backends/compiler.py

```diff
@@ -8,6 +9,101 @@
 from types import ModuleType
 
 
+# This class handles the properties for the given function parameters
+# Different backends can add more properties to the common ones
+class AttrsDescriptor:
+    arg_properties: dict[str, list]
+    property_val: dict[str, int]
+
+    def __init__(self, params=None, args=None):
+        self.arg_properties = {}
+        self.property_val = {"tt.divisibility": 16, "tt.equal_to_1": 1}
+
+        if (params is None) or (args is None):
+            return
+
+        assert (len(params) == len(args))
+
+        # Divisibility property
+        self.arg_properties["tt.divisibility"] = [
+            param.num for param, arg in zip(params, args) if AttrsDescriptor.is_divisible_by_16(arg)
+            and not param.do_not_specialize and not param.do_not_specialize_on_alignment
+        ]
+
+        # Equal to 1 property
+        self.arg_properties["tt.equal_to_1"] = [
+            param.num for param, arg in zip(params, args) if AttrsDescriptor.is_1(arg) and not param.do_not_specialize
+        ]
+
+    # Get the function attributes as a dict like:
+    # {
+    #   "arg0" : [(prop_name00, val00), (prop_name01, val01), ...)]}
+    #   "arg1" : [(prop_name10, val10), (prop_name11, val11), ...)]}
+    # }
+    def get_fn_attrs(self):
+        attrs = {}
+        for prop_name, arg_set in self.arg_properties.items():
+            for arg in arg_set:
+                attrs[arg] = attrs.get(arg, []) + [(prop_name, 16)]
+        return attrs
+
+    # Return the same object, without the given attribute `attr_name`
+    def erase_property(self, attr_name):
+        import copy
+        c = copy.deepcopy(self)
+        if attr_name in c.arg_properties:
+            c.arg_properties.pop(attr_name)
+        return c
+
+    def __getitem__(self, attr_name):
+        if attr_name in self.arg_properties:
+            return self.arg_properties[attr_name]
+        return None
+
+    def hash(self):
+        key = str([sorted(x) for x in self.__dict__.values()])
+        return hashlib.sha256(key.encode("utf-8")).hexdigest()
+
+    def to_dict(self):
+        return self.arg_properties
+
+    # Create the class from a set of hints that are passed out. So instead
+    # of calling the specializations, it's the user that tells us what property
+    # each argument has
+    @staticmethod
+    def from_hints(hints):
+        attrsDescriptor = AttrsDescriptor()
+        for prop_name, prop_val in attrsDescriptor.property_val.items():
+            attrsDescriptor.arg_properties[prop_name] = [i for i, h in hints.items() if h == prop_val]
```

Comment:
Hmm, `hints` should differentiate between divisble by 16 and equal to 1?

Reply:
Why? I mean, `hints` is a list like `[(0, 1), (1,16), (2, 16), (3,1)....]`. So the `AttrsDescriptor` can parse this list and form the different properties.

Reply:
Okay I guess I was just quite confused about the hints format. The docstring helps to clarify here. Thanks.